### PR TITLE
Upgrade to dazzle v2 & use Github Actions for build

### LINKED ISSUE #574: Release dazzle v2 for workspace-full
Swap to v2 image for workspace-full
- [x] move v1 image content to a legacy sub-folder
- [x] move v2 image content to the root
- [x] remove CircleCI
- [x] Resolve the conflict for full/Dockerfile
- [x] Confirm test success for `tool-postgresql` [in this action](https://github.com/gitpod-io/workspace-images/actions/runs/1737842982), potentially fixed in 645753afa6d043e244c97dc4378b8bd7100177f6 
- [x] Investigate nature of Python issues with full image causing minor changes in dev image - is there a solve in the full we can do?
- [x] Squash #593 to less commits :wink: 
- [x] measure usage of new workspace-full to assess proper dog fooding was done (SQL query in DB)
- [x] dog food internally, and handle issues with new workspace-full
- [x] update workspace-full repo with a "escape hatch" text, to help users who have trouble with new image
- [ ] push new workspace-full to Docker Hub with latest tag

### PR DESCRIPTION
**Pull request by princerachit**
## Description

Use dazzle v2 to create images that are published from this repo.
Introduce GH Actions in order to build the images and push them to and internal GAR.
Remove Circle CI integration.

## Related Issue(s)

Fixes #574 

## How to test

The PR already includes a standard set of tests to validate the changes going into each image. A green GH action proves that test ran successfully.
We have additionally used the generated image in the gitpod-io/gitpod repo to test the worksapce full image.

No manual testing is required.

## Release Notes

```release-note
Use dazzle v2 to build workspace images
```

### CODE REVIEW

File: dazzle-rewrite/dazzle.yaml [hunk 1]
```diff
@@ -1,36 +1,58 @@
...
-    #   - essentials
-    #   chunks:
-    #   - lang-c
-    # - name: lang-clojure
-    #   ref:
-    #   - essentials
-    #   chunks:
-    #   - lang-clojure
+      chunks:
+        - tool-docker
```
Comment (kylos101) [inline dazzle-rewrite/dazzle.yaml:5]: @aledbf What other tool chunks should go in base? I am thinking `tool-tailscale` too, but nothing else.

File: dazzle-rewrite/dazzle.yaml [hunk 2]
```diff
@@ -1,36 +1,58 @@
...
+    - name: go
+      chunks:
+        - lang-go:1.17.5
+    - name: node
+      chunks:
+        - lang-node:16.13.1
+    - name: node-lts
+      chunks:
+        - lang-node:16.13.1
+    - name: python
```
Comment (kylos101) [inline dazzle-rewrite/dazzle.yaml:44]: Good call adding versions for Go, Java, Node, and Ruby. :smiley: 

[We'll want to add versions for Python too.](https://github.com/gitpod-io/workspace-images/issues/600) :snake:

File: dazzle-rewrite/chunks/tool-docker/Dockerfile [hunk 3]
```diff
@@ -12,7 +12,7 @@ RUN curl -o /var/lib/apt/dazzle-marks/docker.gpg -fsSL https://download.docker.c
 RUN curl -o /usr/bin/slirp4netns -fsSL https://github.com/rootless-containers/slirp4netns/releases/download/v1.1.12/slirp4netns-$(uname -m) \
     && chmod +x /usr/bin/slirp4netns
 
-RUN curl -o /usr/local/bin/docker-compose -fsSL https://github.com/docker/compose/releases/download/1.29.2/docker-compose-Linux-x86_64 \
+RUN curl -o /usr/local/bin/docker-compose -fsSL https://github.com/docker/compose/releases/download/v2.2.2/docker-compose-linux-$(uname -m) \
```
Comment (kylos101) [inline dazzle-rewrite/chunks/tool-docker/Dockerfile:15]: We should add a related test for the version of docker-compose.

File: dazzle-rewrite/chunks/lang-ruby/Dockerfile [hunk 4]
```diff
@@ -1,15 +1,17 @@
...
 ENV TRIGGER_REBUILD=1
 RUN curl -fsSL https://rvm.io/mpapis.asc | gpg --import - \
     && curl -fsSL https://rvm.io/pkuczynski.asc | gpg --import - \
     && curl -fsSL https://get.rvm.io | bash -s stable \
     && bash -lc " \
         rvm requirements \
-        && rvm install 2.7.4 \
-        && rvm use 2.7.4 --default \
+        && rvm install ${RUBY_VERSION} \
+        && rvm use ${RUBY_VERSION} --default \
```
Comment (kylos101) [inline dazzle-rewrite/chunks/lang-ruby/Dockerfile:14]: We should write a test that inspects the Ruby version.

File: dazzle-rewrite/chunks/lang-node/Dockerfile [hunk 5]
```diff
@@ -8,8 +8,8 @@ ENV NODE_VERSION=${NODE_VERSION}
 ENV TRIGGER_REBUILD=2
 RUN curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.0/install.sh | PROFILE=/dev/null bash \
     && bash -c ". .nvm/nvm.sh \
-        && nvm install $NODE_VERSION \
-        && nvm alias default $NODE_VERSION \
+        && nvm install v${NODE_VERSION} \
+        && nvm alias default v${NODE_VERSION} \
```
Comment (kylos101) [inline dazzle-rewrite/chunks/lang-node/Dockerfile:12]: We should write a test that inspects the version of node.

File: dazzle-rewrite/chunks/lang-java/Dockerfile [hunk 6]
```diff
@@ -1,11 +1,13 @@
...
 FROM ${base}
 
+ARG JAVA_VERSION
+
 ENV TRIGGER_REBUILD=1
 USER gitpod
 RUN curl -fsSL "https://get.sdkman.io" | bash \
  && bash -c ". /home/gitpod/.sdkman/bin/sdkman-init.sh \
-             && sdk install java 11.0.13.fx-zulu \
+             && sdk install java ${JAVA_VERSION}.fx-zulu \
```
Comment (kylos101) [inline dazzle-rewrite/chunks/lang-java/Dockerfile:10]: We should write a test that inspects the version of Java

File: dazzle-rewrite/chunks/lang-go/chunk.yaml [hunk 7]
```diff
@@ -1,4 +1,19 @@
 variants:
+  - name: "1.17"
```
Comment (kylos101) [inline dazzle-rewrite/chunks/lang-go/chunk.yaml:2]: We should write a test that inspects the version of Go.

**Review by kylos101 — COMMENTED** [review]
@aledbf you rock! I like the addition of cache in Github actions. :smiley: 

There are 3 comments I specifically addressed to you @aledbf . LMKWYT?

The balance of comments are for this effort in general. You're welcome to handle too, but @princerachit  and I can also do when we return.

File: .github/workflows/pull-request.yml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+          repository: gitpod-io/workspace-images
+
+      - name: 🔆 Install dazzle
+        run: |
+          curl -sSL https://github.com/gitpod-io/dazzle/releases/download/v0.1.6/dazzle_0.1.6_Linux_x86_64.tar.gz | sudo tar -xvz -C /usr/local/bin
+
+      - name: 🏗️ Setup buildkit
+        run: |
+          curl -sSL https://github.com/moby/buildkit/releases/download/v0.9.3/buildkit-v0.9.3.linux-amd64.tar.gz | sudo tar xvz -C /usr
+          sudo buildkitd --oci-worker=true --oci-worker-net=host --debug --group docker &
```
[the remark this answers is not in the document]
Comment (aledbf) [reply .github/workflows/pull-request.yml:21]: Just to mirror what we use in [image-builder-bob](https://github.com/gitpod-io/gitpod/blob/master/components/image-builder-bob/pkg/builder/builder.go#L272)

File: .github/sync-containers.yml [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# https://github.com/containers/skopeo/blob/main/docs/skopeo-sync.1.md#yaml-file-content-used-source-for---src-yaml
+localhost:5000:
+  images:
+    workspace-base: []
+    workspace-full:
+      - "1.0"
+      - "2.0"
+      - "sha256:0000000000000000000000000000000011111111111111111111111111111111"
```
[the remark this answers is not in the document]
Comment (aledbf) [reply .github/sync-containers.yml:8]: This was just an example, not what we should use. At this point, I realized I drastically changed the approach.

> Do we need to add more tags? If we were to add more tags, like human friendly / semvar tag:

Yes, that the idea ;)

File: dazzle-rewrite/dazzle.yaml (hunk 1 above)
Comment (aledbf) [reply dazzle-rewrite/dazzle.yaml:5]: > What other tool chunks should go in base?

Good question. I put docker because is what makes sense.

> I am thinking tool-tailscale too, but nothing else.

Sure?

File: .github/workflows/push-main.yml [hunk 10]
```diff
@@ -83,11 +84,11 @@ jobs:
...
-          echo "${{ steps.auth.outputs.access_token }}" | skopeo login -u oauth2accesstoken --password-stdin https://europe-docker.pkg.dev
+          echo "${{ steps.auth.outputs.access_token }}" | skopeo login -u oauth2accesstoken --password-stdin https://${{env.IMAGE_REPO}}
 
       - name: 🐳 Sync images to Artifact Registry
         run: |
           skopeo sync \
             --src yaml \
             --dest docker \
-            .github/sync-containers.yml europe-docker.pkg.dev/workspace-images
+            .github/sync-containers.yml ${{env.IMAGE_REPO}}/workspace-base-images/workspace-base-images/workspace-image
```
Comment (aledbf) [inline .github/workflows/push-main.yml:94]: why `workspace-base-images/workspace-base-images/workspace-image` ?

File: .vscode/launch.json [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            "request": "launch",
+            "mode": "auto",
+            "program": "${fileWorkspaceFolder}/dazzle",
+            "args": [
+                "build",
+                "--addr",
+                "unix:///run/buildkit/buildkitd.sock",
+                "--context",
+                "${fileWorkspaceFolder}/dazzle-rewrite/",
+                "europe-docker.pkg.dev/prince-tf-experiments/dazzle/workspace-images",
```
Comment (kylos101) [inline .vscode/launch.json:27]: @princerachit Can you change this to our new workspace-base-images registry?

File: .github/workflows/push-main.yml (hunk 10 above)
Comment (kylos101) [reply .github/workflows/push-main.yml:94]: Just temporary.  I want to test skopeo accessing GAR, before rebuilding GAR to have a nicer/cleaner URL. This is the URL we used earlier.

CC: @princerachit

Comment (kylos101) [reply .github/workflows/push-main.yml:94]: We'll rebuild GAR and clean-up the URL before merging. 

CC: @princerachit

File: .github/workflows/pull-request.yml [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+          key: ${{ runner.os }}-pull-request-cache
+
+      - name: 📦 Setup local registry
+        run: |
+          docker run -it --detach --publish 5000:5000 --volume ~/registry:/var/lib/registry registry:2
+
+      - name: 🔨 Dazzle build
+        run: |
+          cd dazzle-rewrite
+          dazzle build localhost:5000/workspace-images
```
Comment (aledbf) [inline .github/workflows/pull-request.yml:38]: Please add the flag `--chunked-without-hash` (@csweichel suggestion)

File: .github/workflows/push-main.yml [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+          key: ${{ runner.os }}-main-cache
+
+      - name: 📦 Setup local registry
+        run: |
+          docker run -it --detach --publish 5000:5000 --volume ~/registry:/var/lib/registry registry:2
+
+      - name: 🔨 Dazzle build
+        run: |
+          cd dazzle-rewrite
+          dazzle build localhost:5000/workspace-images
```
Comment (aledbf) [inline .github/workflows/push-main.yml:65]: Please add the flag `--chunked-without-hash`

File: .github/workflows/push-main.yml (hunk 10 above)
Comment (kylos101) [reply .github/workflows/push-main.yml:94]: For posterity, I believe the path must be three layers deep. 

When we rebuild GAR, it could look like this:
1. Project -> **gitpod-artifacts**
2. Org -> **docker-dev** (to denote the registry type and general purpose)
3. Repository -> **workspace-base-images** (to denote what this specific repo will contain)

File: .github/workflows/push-main.yml [hunk 14]
```diff
@@ -83,11 +84,11 @@ jobs:
 
       - name: Configuring container registry credentials
         run: |
-          echo "${{ steps.auth.outputs.access_token }}" | skopeo login -u oauth2accesstoken --password-stdin https://europe-docker.pkg.dev
+          echo "${{ steps.auth.outputs.access_token }}" | skopeo login -u oauth2accesstoken --password-stdin https://${{env.IMAGE_REPO}}
```
Comment (kylos101) [inline .github/workflows/push-main.yml:87]: @aledbf can you think of another way to login with skopeo w/o echoing the access token? I know it's only valid for an hour, I'm just trying to avoid it being potentially output to logs and used.

File: dazzle-rewrite/chunks/lang-go/chunk.yaml (hunk 7 above)
Comment (princerachit) [reply dazzle-rewrite/chunks/lang-go/chunk.yaml:2]: Every chunk has only one test file. So it is tricky to verify the version. What we can do is to test if the version is one of `x,y,z,a,b,c...` variants.

File: .github/workflows/push-main.yml (hunk 10 above)
Comment (princerachit) [reply .github/workflows/push-main.yml:94]: @aledbf explaining further what @kylos101 added `workspace-base-images(1)/workspace-base-images(2)/workspace-image(3)`.

(1) Project Name
(2) Repo Name
(3) Image name

File: dazzle-rewrite/chunks/lang-go/chunk.yaml (hunk 7 above)
Comment (princerachit) [reply dazzle-rewrite/chunks/lang-go/chunk.yaml:2]: Added in a new commit

File: .github/workflows/push-main.yml [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+on:
+  push:
+    branches:
+      - ws-team/dazzle-rewrite
+
+jobs:
+  # Sync container images between the local registry and GCP Artifact registry.
+  # To implementr manual approvals, the workflow uses an Environment.
+  #
+  # From your GitHub repo clock Settings. In the left menu, click Environments.
```
[the remark this answers is not in the document]
Comment (kylos101) [reply .github/workflows/push-main.yml:12]: This is currently disabled, I just enabled it and allowed `gitpod-io/engineering-workspace` for `production`.

File: .github/workflows/pull-request.yml (hunk 12 above)
Comment (kylos101) [reply .github/workflows/pull-request.yml:50]: @aledbf @csweichel @princerachit 

A side effect of doing two builds, is that each build takes an hour now. :thinking: [[1](https://github.com/gitpod-io/workspace-images/actions/runs/1653880874)][[2](https://github.com/gitpod-io/workspace-images/actions/runs/1653974963)]

For pull requests, I think it may be helpful to omit `--chunked-without-hash`, because then we can leverage the cache. 

For main, we need `--chunked-without-hash`. We can disable running of tests if `--chunked-without-hash` is specfieid (a dazzle code change because no flag exists), knowing that the next `dazzle build` will run the chunk tests. 

WDYT?

File: dazzle-rewrite/base/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+        jq \
+        less \
+        locales \
+        man-db \
+        nano \
+        ripgrep \
+        software-properties-common \
+        sudo \
+        time \
+        emacs-nox \
```
[the remark this answers is not in the document]
Comment (kylos101) [reply dazzle-rewrite/base/Dockerfile:25]: [Found it](https://github.com/gitpod-io/workspace-images/commit/5a06e6a16d202988f4a2d6e08b9a81b9aee65451#diff-b093d34eec92adf6e0cc6e445edb7147ea45d25fb0c98f2f84a60f5aae6f34f7)

Whatcha think about continued use?

File: dazzle-rewrite/dazzle.yaml (hunk 1 above)
Comment (princerachit) [reply dazzle-rewrite/dazzle.yaml:5]: added tailscale

File: .github/workflows/push-main.yml (hunk 10 above)
Comment (princerachit) [reply .github/workflows/push-main.yml:108]: resolving this comment as we have a [separate issue](https://github.com/gitpod-io/workspace-images/issues/539) to track this

File: .github/workflows/pull-request.yml (hunk 12 above)
Comment (princerachit) [reply .github/workflows/pull-request.yml:49]: >For pull requests, I think it may be helpful to omit --chunked-without-hash, because then we can leverage the cache

Sounds good to me. Let me get rid of this for PRs

>For main, we need --chunked-without-hash. We can disable running of tests if --chunked-without-hash is specfieid (a dazzle code change because no flag exists), knowing that the next dazzle build will run the chunk tests.

👍🏾

File: dazzle-rewrite/chunks/lang-node/chunk.yaml [hunk 17]
```diff
@@ -0,0 +1,10 @@
+variants:
+  - name: "16.13.0"
+    args:
+      NODE_VERSION: 16.13.0
+  - name: "16.13.1"
+    args:
+      NODE_VERSION: 16.13.1
+  - name: "17"
```
[the remark this answers is not in the document]
Comment (princerachit) [reply dazzle-rewrite/chunks/lang-node/chunk.yaml:8]: What are the downside of using hashes with chunked image? Is it the build time or users needing to pull a new chunk because a new hash will be generated when rebuilding?

File: .github/workflows/pull-request.yml (hunk 12 above)
Comment (princerachit) [reply .github/workflows/pull-request.yml:59]: I am resolving this conversation, feel free to comment again if you don't agree with changes

File: .github/workflows/push-main.yml (hunk 13 above)
Comment (princerachit) [reply .github/workflows/push-main.yml:87]: done

**Comment by gengjiawen** [comment]
Is this closed to produciton ?

File: dazzle-rewrite/chunks/lang-c/Dockerfile [hunk 18]
```diff
@@ -2,11 +2,14 @@ ARG base
 FROM ${base}
 
 USER root
+
 # Dazzle does not rebuild a layer until one of its lines are changed. Increase this counter to rebuild this layer.
-ENV TRIGGER_REBUILD=1
-RUN curl -o /var/lib/apt/dazzle-marks/llvm.gpg -fsSL https://apt.llvm.org/llvm-snapshot.gpg.key \
-    && apt-key add /var/lib/apt/dazzle-marks/llvm.gpg \
-    && echo "deb https://apt.llvm.org/focal/ llvm-toolchain-focal main" >> /etc/apt/sources.list.d/llvm.list \
+ENV TRIGGER_REBUILD=3
```
Comment (princerachit) [inline dazzle-rewrite/chunks/lang-c/Dockerfile:7]: did we need this change? The build gets retriggered in the beow step due to change in the step

**Comment by kylos101** [comment]
@princerachit I unlinked this PR from the epic (we don't want the epic to be closed when this PR closes, which is what our workflow would have done). Also, I linked it to #574, and updated the tasks in #574 to reflect what is left. May I ask you to peek at the tasks in #574 for Monday? I'll review this PR again tomorrow with a fresh pair of eyes.

**Review by kylos101 — APPROVED** [review]
/LGTM

File: dazzle.yaml [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+      - base
+      chunks:
+        - lang-rust
+    - name: dotnet-vnc
+      ref:
+      - base
+      chunks:
+        - tool-vnc
+        - tool-dotnet
+    - name: postgresql
```
Comment (tnir) [inline dazzle.yaml:79]: @princerachit Why did you rename `gitpod/workspace-postgres` to `gitpod/workspace-postgresql`? cc @kylos101
